Initial commit

cap10morgan opened the pull request:
I used this to start election-notification-http-api, which will be up in a PR as soon as I verify that it works in the krakenstein.

cap10morgan committed:
[WM] rename kraken_works.clj to kraken_http_api.clj

cap10morgan committed:
[WM] rename resources/leiningen/new subdir

cap10morgan committed:
[WM] update project name, desc, and url

cap10morgan committed:
[WM] remove profiles.clj.sample

These typically won't need Datomic access, which is what we use profiles.clj for.

cap10morgan committed:
[WM] install server & service ns'es instead of core

cap10morgan committed:
[WM] rename core.clj to server.clj

cap10morgan committed:
[WM] update deps; remove pedestal-tb snapshot

cap10morgan committed:
[WM] update lein profiles

cap10morgan committed:
[WM] update queue to kehaar 0.5.0

cap10morgan committed:
[WM] convert server ns to pedestal

cap10morgan committed:
[WM] add basic service ns

cap10morgan committed:
[WM] close ok-* channels

cap10morgan committed:
[WM] datomic out, pedestal in (config.edn)

cap10morgan committed:
[WM] remove datomic from docker-compose.yml

cap10morgan committed:
[WM] remove profiles.clj COPY from Dockerfile

cap10morgan committed:
[WM] add logback config

cap10morgan committed:
[WM] remove datomic repo from project.clj

cap10morgan committed:
[WM] fix weird core.async behavior w/ pedestal

Before this change, I was seeing a ClassNotFoundException when it was compiling kehaar.wire-up. Removing pedestal made it go away, but so did removing the try-catch from kehaar.core/go-handler.

I *think* the much-newer version of tools.analyzer.jvm that pedestal brings in (versus the one core.async depends on) was to blame. Hopefully this won't break when core.async updates to a newer version...

cap10morgan committed:
[WM] remove some unused requires from queue

cap10morgan committed:
[WM] pass rabbit config to connect-with-retries

cap10morgan committed:
[WM] start responder for ok handler

cap10morgan committed:
[WM] remove README reference to profiles.clj

cap10morgan committed:
[WM] pass server config to start-http-server

cap10morgan committed:
[WM] convert handlers-test to service-test

Test the /ping endpoint to show how it's done.

cap10morgan committed:
[WM] add logback.xml to template renderer

cap10morgan committed:
[WM] fix git command in lein new output

-a after `git add .` is redundant (and potentially not what you want if you do anything to the working directory between the 2 commands).

cap10morgan committed:
[WM] fix indentation of service-test requires

...and add core.async since you'll definitely need that to test async interceptors.

cap10morgan commented:
@tank157 Do you think you'll have time to look at this today? Would a walkthrough help? I can assign to someone else if you're swamped.

tank157 commented:
@cap10morgan I should have time to go over it later today, but if there's a rush, go ahead and assign it to someone else.

cap10morgan commented:
@tank157 Today is fine, I just wanted to leave time to address any issues before I go on vacation. Thanks!

tank157 commented on resources/leiningen/new/kraken_http_api/service.clj:
```diff
@@ -0,0 +1,35 @@
+            [kehaar.core :as k]
+            [clojure.core.async :refer [chan go alt! timeout]]))
+
+(def ping
```
Just a thought, but should the ping make use of the ok queue? I mean, it's of marginal value to know the http-api is up, it's more important to know that it's up and able to talk with the back end. Or is there a better way to do that?

cap10morgan replied:
I think it's pretty useful even just knowing that the HTTP server is up and running.

The thing that seems weird about doing anything with the :ok queue is that it would be a message out and back into the same component. That's perfectly valid, I suppose, but I'm not sure it tells you very much.

Or did you mean send a message to the backend *-works component's :ok queue and wait for a response? The weird thing there is that these HTTP API gateways are specifically intended _not_ to be tied to 1:1 relationships with backend components. They're intended to implement useful, predictable, pragmatic HTTP APIs for us and other developers and talk to whatever backend components they need to to do that. It just so happens that all the ones we've built so far only talk to one backend component each. If that ends up being true everywhere, then maybe we can bake it in further. But it feels too early to start doing that here.

Or I suppose we could gather a list of every backend component that they talk to and send an :ok message to each of them in ping... Hmm...

There could be something useful here, I'm just not sure what it should look like at this stage. Thoughts?

tank157 replied:
It's not something we need to solve now by any means. And maybe ping is fine to just stay there as a "http server is up" endpoint, but it would be nice to also have a way of knowing that this http server can actually serve it's purpose by talking with backends.

tank157 commented:
Other than the one comment, LGTM

cap10morgan committed:
[WM] use linear-search pedestal router

This was what pedestal < 0.4.0 used and it is much more flexible and
predictable than the new default prefix-tree router in 0.4.0. The main
advantage of prefix-tree is that it is more performant with lots of
routes. But our very architecture makes it unlikely we'll have enough
routes to matter in any one -http-api component.
